### PR TITLE
 Reaction Split Transformer

### PR DESCRIPTION
**Pull request by Suzukazole**
## Description
Adds the transformer that splits the reaction SMILES into the source and target strings required for a translation setting. 

TODO:
- [x] Add transformer
- [x] Add option to perform reagent mixing(extends the loader) 
- [x] Update documentation
- [x] Add tests for separation and mixing.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [X] Documentations (modification for documents)

## Checklist

- [X] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [X] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.22.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [X] Run `flake8 <modified file> --count` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/trans/transformers.py [hunk 1]
```diff
@@ -2469,3 +2469,62 @@ def median_filter(self, size):
     image = Image.fromarray(self.Image)
     image = image.filter(ImageFilter.MedianFilter(size=size))
     return np.array(image)
+
+
+class RxnSplitTransformer(Transformer):
```
Comment (rbharath) [inline deepchem/trans/transformers.py:2474]: Make sure to add a docstring here with a usage example

File: deepchem/trans/transformers.py [hunk 2]
```diff
@@ -2469,3 +2469,62 @@ def median_filter(self, size):
     image = Image.fromarray(self.Image)
     image = image.filter(ImageFilter.MedianFilter(size=size))
     return np.array(image)
+
+
```
Comment (rbharath) [inline deepchem/trans/transformers.py:2473]: Make sure to add the new transformer to the transformer docs and to add some unit tests

**Review by rbharath — COMMENTED** [review]
Took a first look and made some preliminary comments

File: deepchem/trans/transformers.py (hunk 1 above)
Comment (Suzukazole) [reply deepchem/trans/transformers.py:2474]: yep! I will add those. The input is the reaction SMILES in the reactant>reagent>product format, and can also handle the case when there are no reagents i.e reactant>>product format.

File: deepchem/trans/transformers.py (hunk 2 above)
Comment (Suzukazole) [reply deepchem/trans/transformers.py:2473]: Will do!

File: deepchem/trans/tests/test_rxn_transform.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+#        'CC(C)CC(=O)c1ccc(O)nc1'],
+#       ['CN.O=C(O)c1ccc(Cl)c([N+](=O)[O-])c1>O',
+#        'CNc1ccc(C(=O)O)cc1[N+](=O)[O-]'],
+#       ['CCn1cc(C(=O)O)c(=O)c2cc(F)c(-c3ccc(N)cc3)cc21.O=CO>',
+#        'CCn1cc(C(=O)O)c(=O)c2cc(F)c(-c3ccc(NC=O)cc3)cc21']], dtype='<U51'), [], [], [])
+
+
+class TestRxnSplitTransformer(unittest.TestCase):
+  """Tests the Reaction split transformer for the source/target splitting and
+    for the reagent mixing operation.
```
Comment (rbharath) [inline deepchem/trans/tests/test_rxn_transform.py:24]: Make sure the docstring is formatted correctly. yapf sometimes messes this up so you might need to fix manually

File: deepchem/trans/tests/test_rxn_transform.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    """
+
+  def test_split(self):
+    """
+        Tests the source/target split from an input reaction SMILES.
```
Comment (rbharath) [inline deepchem/trans/tests/test_rxn_transform.py:29]: Docstring formatting issue here too

File: deepchem/trans/transformers.py [hunk 5]
```diff
@@ -2469,3 +2469,120 @@ def median_filter(self, size):
...
+  The input is expected to be in the form reactant>reagent>product. The source
+  string would be reactants>reagents and the target string would be the products.
+
+  The transformer can also separate the reagents from the reactants for a mixed
+  training mode. This can be toggled (default True) by setting the value of
+  sep_reagent while calling the transformer.
+
+  Examples
+  --------
+  >>> # When mixed training is toggled.
```
Comment (rbharath) [inline deepchem/trans/transformers.py:2487]: What is mixed training? Can you add more text above that explains the concept?

File: deepchem/trans/transformers.py [hunk 6]
```diff
@@ -2469,3 +2469,120 @@ def median_filter(self, size):
...
+  >>> trans = RxnSplitTransformer(sep_reagent=True)
+  >>> split_reactions = trans.transform_array(X=reactions, y=np.array([]), w=np.array([]), ids=np.array([]))
+  >>> split_reactions
+  (array([['CC(C)C[Mg+].CON(C)C(=O)c1ccc(O)nc1>C1CCOC1.[Cl-]',
+           'CC(C)CC(=O)c1ccc(O)nc1'],
+          ['CN.O=C(O)c1ccc(Cl)c([N+](=O)[O-])c1>O',
+           'CNc1ccc(C(=O)O)cc1[N+](=O)[O-]'],
+          ['CCn1cc(C(=O)O)c(=O)c2cc(F)c(-c3ccc(N)cc3)cc21.O=CO>',
+           'CCn1cc(C(=O)O)c(=O)c2cc(F)c(-c3ccc(NC=O)cc3)cc21']], dtype='<U51'), array([], dtype=float64), array([], dtype=float64), array([], dtype=float64))
+  >>> # When mixed training is diabled.
```
Comment (rbharath) [inline deepchem/trans/transformers.py:2506]: Just do a text break like
```
When mixed training is disabled, you get the following outputs

>>> trans_disable = ...
```

File: deepchem/trans/transformers.py [hunk 7]
```diff
@@ -2469,3 +2469,120 @@ def median_filter(self, size):
...
+  sep_reagent while calling the transformer.
+
+  Examples
+  --------
+  >>> # When mixed training is toggled.
+  >>> import numpy as np
+  >>> from deepchem.trans.transformers import RxnSplitTransformer
+  >>> reactions = np.array(
+    [
+        "CC(C)C[Mg+].CON(C)C(=O)c1ccc(O)nc1>C1CCOC1.[Cl-]>CC(C)CC(=O)c1ccc(O)nc1",
```
Comment (rbharath) [inline deepchem/trans/transformers.py:2492]: Can you simplify this to have only 2 reactions instead of 3? Will make it easier to read

**Review by rbharath — COMMENTED** [review]
Just reviewed! A few issues mostly around the docstrings that should be easy to resolve hopefully

File: deepchem/trans/tests/test_rxn_transform.py (hunk 3 above)
Comment (Suzukazole) [reply deepchem/trans/tests/test_rxn_transform.py:24]: Alright, I will fix that.

File: deepchem/trans/transformers.py (hunk 7 above)
Comment (Suzukazole) [reply deepchem/trans/transformers.py:2492]: Yup, I will do that. Only 2 are needed to demonstrate the mixed training functionality.

File: deepchem/trans/transformers.py (hunk 5 above)
Comment (Suzukazole) [reply deepchem/trans/transformers.py:2487]: It just refers to the fact that the reagents and reactants are no longer separated by a '>'. I will clarify this on the doc as well!

**Review by rbharath — APPROVED** [review]
Marking this as good to go from my end! Will merge in on Monday in case there aren't any further comments

File: deepchem/trans/tests/test_rxn_transform.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
+import unittest
+import numpy as np
+
+from deepchem.trans.transformers import RxnSplitTransformer
+
+reactions: np.ndarray = np.array(
```
[the remark this answers is not in the document]
Comment (vsomnath) [reply deepchem/trans/tests/test_rxn_transform.py:6]: Not sure what the current convention is on testing, but in the past, I remember seeing a  `setUp` method for the `TestXYZ` class. @rbharath: This one and the old style are fine by me though.

**Pull request merged**